modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/block/BlockRelations.scala

```diff
@@ -0,0 +1,40 @@
+package org.tessellation.dag.l1.domain.block
+
+import cats.effect.kernel.Async
+import cats.syntax.applicative._
+import cats.syntax.eq._
+import cats.syntax.functor._
+import cats.syntax.traverse._
+
+import org.tessellation.dag.domain.block.DAGBlock
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.BlockReference
+import org.tessellation.schema.transaction.TransactionReference
+import org.tessellation.security.Hashed
+import org.tessellation.security.signature.Signed
+
+object BlockRelations {
+
+  def dependsOn[F[_]: Async: KryoSerializer](
+    blockToTestAgainst: Hashed[DAGBlock]
+  )(block: Signed[DAGBlock]): F[Boolean] = dependsOn(Set(blockToTestAgainst))(block)
+
+  def dependsOn[F[_]: Async: KryoSerializer](
+    blocksToTestAgainst: Set[Hashed[DAGBlock]],
+    blocksRefsToTestAgainst: Set[BlockReference] = Set.empty
+  )(block: Signed[DAGBlock]): F[Boolean] = {
+    def dstAddresses = blocksToTestAgainst.flatMap(_.transactions.toSortedSet.toList.map(_.value.destination))
+
+    def isChild =
+      block.parent.exists(parentRef => (blocksToTestAgainst.map(_.ownReference) ++ blocksRefsToTestAgainst).exists(_ === parentRef))
+    def hasReferencedAddress = block.transactions.map(_.source).exists(srcAddress => dstAddresses.exists(_ === srcAddress))
+    def hasReferencedTx = blocksToTestAgainst.toList
+      .flatTraverse(_.transactions.toSortedSet.toList.traverse(TransactionReference.of(_)))
+      .map(_.toSet)
+      .map { txRefs =>
+        block.transactions.map(_.parent).exists(txnParentRef => txRefs.exists(_ === txnParentRef))
+      }
+
+    if (isChild || hasReferencedAddress) true.pure[F] else hasReferencedTx
```

Comment:
You can lift the condition itself into `F` so you will avoid redundant `if (true) then true else hasReferencedTx`

```scala
(isChild || hasReferencedAddress).pure[F].orElse(hasReferencedTx)
```

Reply:
Changed.

Reply:
In fact above is not correct as `orElse` in this case is called when error is thrown not when value `false`. I haven't found any other type class with `orElse` and instance for `Boolean`. I have reverted it to previous version.

Reply:
Maybe just like this?
`hasReferencedTx.map(_ || isChild || hasReferencedAddress)`

Reply:
@TheMMaciek nope. `isChild` and `hasReferencedAddress` are simple checks. `hasReferencedTx` requires building refs, so it should be lazy evaluated. that being said, `if` condition written by @kriscieslak  looks fine for me.

Reply:
Ok, there might be some performance gain from doing it that way.

Reply:
I guess this would work, but it's not even shorter than the original so not sure we want it 😄 
```
(isChild.pure[F], hasReferencedAddress.pure[F], hasReferencedTx).existsM(identity)
```

Reply:
Ok, then I am fine with the original solution.